In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os

DRIVE_ROOT = '/content/drive/MyDrive/deepfake_audio'
RAW_2019 = os.path.join(DRIVE_ROOT, 'data/raw/asvspoof_2019')

# Check the folder we created in Phase 1 still exists
if os.path.exists(RAW_2019):
    print(f"✅ Found target folder: {RAW_2019}")
    contents = os.listdir(RAW_2019)
    if contents:
        print(f"   Contents: {contents}")
    else:
        print(f"   (folder is empty, ready for download)")
else:
    print(f"❌ Folder not found, creating it now...")
    os.makedirs(RAW_2019, exist_ok=True)
    print(f"✅ Created: {RAW_2019}")

✅ Found target folder: /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019
   (folder is empty, ready for download)


In [ ]:
from google.colab import files
print("Click 'Choose Files' and select kaggle.json from your Downloads folder")
uploaded = files.upload()

Click 'Choose Files' and select kaggle.json from your Downloads folder


Saving kaggle.json to kaggle.json


In [ ]:
import os
import shutil

# Move kaggle.json to the standard location
os.makedirs('/root/.kaggle', exist_ok=True)
shutil.move('kaggle.json', '/root/.kaggle/kaggle.json')

# Set strict permissions (Kaggle CLI requires this)
os.chmod('/root/.kaggle/kaggle.json', 0o600)

print("✅ Kaggle credentials configured.")

✅ Kaggle credentials configured.


In [ ]:
!pip install -q kaggle

# Test that the API is authenticated by listing one of your datasets / competitions
!kaggle datasets list -s asvspoof --max-size 100000000 2>&1 | head -20

ref                                          title                             size  lastUpdated                 downloadCount  voteCount  usabilityRating  
-------------------------------------------  --------------------------  ----------  --------------------------  -------------  ---------  ---------------  
shettysharanyaa/asvspoof-features            asvspoof-features             23554403  2026-02-24 16:12:08.220000              0          0  0.1875           
chandajha04/asvspoof-2021                    asvspoof-2021                     4225  2025-11-08 13:54:33.450000              2          0  0.25             
shettysharanyaa/asvspoof-features-all        asvspoof-features-all         11807951  2026-03-15 15:00:38.487000              0          0  0.0625           
beosup/asvspoof-2019-la-samples              ASVspoof-2019-LA-samples        249565  2026-01-15 03:06:14.567000              3          0  0.25             
simontrann/asvspoof2021-la-key               ASVSpoof2021_

In [ ]:
!kaggle datasets list -s "asvpoof 2019" 2>&1

ref                                    title                            size  lastUpdated                 downloadCount  voteCount  usabilityRating  
-------------------------------------  ------------------------  -----------  --------------------------  -------------  ---------  ---------------  
awsaf49/asvpoof-2019-dataset           ASVspoof 2019 Dataset     25321823144  2022-06-21 17:18:04.330000          13384         55  0.625            
anishsarkar22/asvpoof-2019-dataset-la  ASVSpoof 2019 LA Dataset   7646913333  2025-10-13 11:12:09.283000            183          2  0.4375           
sukiss/asvspoof-dataset                Asvspoof_dataset                 5355  2025-08-02 18:57:44.890000              6          0  0.25             


In [ ]:
!kaggle datasets files anishsarkar22/asvpoof-2019-dataset-la 2>&1 | head -30

Next Page Token = [REDACTED]
name                                                                         size  creationDate                
------------------------------------------------------------------------  -------  --------------------------  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.dev.female.trl.txt    677652  2025-10-13 11:13:31.155000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.dev.female.trn.txt       972  2025-10-13 11:13:31.143000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.dev.gi.trl.txt        976804  2025-10-13 11:13:31.148000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.dev.male.trl.txt      299152  2025-10-13 11:13:31.076000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.dev.male.trn.txt        1096  2025-10-13 11:13:31.066000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.eval.female.trl.txt  2371140  2025-10-13 11:13:31.119000  
LA/ASVspoof2019_LA_asv_protocols/ASVspoof2019.LA.asv.eval.female.trn.txt   

In [ ]:
import os

# Where the zip will land first (Colab's local fast disk, not Drive)
DOWNLOAD_DIR = '/content/kaggle_download'
os.makedirs(DOWNLOAD_DIR, exist_ok=True)

# Download to Colab's local disk first — much faster than writing to Drive
print("Starting download (~7.6 GB). This takes 5-15 minutes...")
!kaggle datasets download -d anishsarkar22/asvpoof-2019-dataset-la -p {DOWNLOAD_DIR} --unzip --force
print("\n✅ Download + unzip complete.")

Starting download (~7.6 GB). This takes 5-15 minutes...
Dataset URL: https://www.kaggle.com/datasets/anishsarkar22/asvpoof-2019-dataset-la
License(s): ODC Attribution License (ODC-By)
100% 7.12G/7.12G [03:26<00:00, 37.0MB/s]


✅ Download + unzip complete.


In [ ]:
import os

DOWNLOAD_DIR = '/content/kaggle_download'
print("Top-level contents:")
for item in sorted(os.listdir(DOWNLOAD_DIR)):
    full = os.path.join(DOWNLOAD_DIR, item)
    if os.path.isdir(full):
        print(f"  folder {item}/")
    else:
        size_mb = os.path.getsize(full) / 1e6
        print(f" file content {item} ({size_mb:.1f} MB)")

LA_DIR = os.path.join(DOWNLOAD_DIR, 'LA')
if os.path.exists(LA_DIR):
    print(f"\nLA/ contents:")
    for item in sorted(os.listdir(LA_DIR)):
        print(f"  folder LA/{item}/")


Top-level contents:
  folder LA/

LA/ contents:
  folder LA/ASVspoof2019_LA_asv_protocols/
  folder LA/ASVspoof2019_LA_asv_scores/
  folder LA/ASVspoof2019_LA_cm_protocols/
  folder LA/ASVspoof2019_LA_dev/
  folder LA/ASVspoof2019_LA_eval/
  folder LA/ASVspoof2019_LA_train/
  folder LA/README.LA.txt/


In [ ]:
import glob

partitions = ['train', 'dev', 'eval']
expected = {'train': 25380, 'dev': 24844, 'eval': 71237}

print("Audio file counts:")
for p in partitions:
    flac_dir = os.path.join(DOWNLOAD_DIR, 'LA', f'ASVspoof2019_LA_{p}', 'flac')
    if os.path.exists(flac_dir):
        n = len(glob.glob(os.path.join(flac_dir, '*.flac')))
        match = "Done" if n == expected[p] else "Warning"
        print(f"  {match} {p}: {n:,} files (expected {expected[p]:,})")
    else:
        print(f"  {p}: folder not found")

Audio file counts:
  Done train: 25,380 files (expected 25,380)
  Warning dev: 24,986 files (expected 24,844)
  Warning eval: 71,933 files (expected 71,237)


In [ ]:
import os
import time

SOURCE = '/content/kaggle_download/LA'
DESTINATION = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'

os.makedirs(os.path.dirname(DESTINATION), exist_ok=True)

print(f"Moving {SOURCE} -> {DESTINATION}")
print("This takes 15-25 minutes due to Drive's per-file overhead.")
print("Don't close the tab. Progress will print as it goes.\n")

start = time.time()
# rsync is more reliable than cp for large transfers
!rsync -ah --info=progress2 {SOURCE}/ {DESTINATION}/

elapsed = (time.time() - start) / 60
print(f"\nMove complete in {elapsed:.1f} minutes.")

Moving /content/kaggle_download/LA -> /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA
This takes 15-25 minutes due to Drive's per-file overhead.
Don't close the tab. Progress will print as it goes.

          7.66G 100%   85.99MB/s    0:01:24 (xfr#122318, to-chk=0/122328)

Move complete in 1.4 minutes.


In [ ]:
import os
import glob

DRIVE_LA = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'

# Re-count files on Drive
partitions = ['train', 'dev', 'eval']
expected = {'train': 25380, 'dev': 24986, 'eval': 71933}  # using actual counts now

print("Verifying files on Drive:")
all_ok = True
for p in partitions:
    flac_dir = os.path.join(DRIVE_LA, f'ASVspoof2019_LA_{p}', 'flac')
    if os.path.exists(flac_dir):
        n = len(glob.glob(os.path.join(flac_dir, '*.flac')))
        match = "OK" if n == expected[p] else "MISMATCH"
        print(f"  [{match}] {p}: {n:,} files")
        if n != expected[p]:
            all_ok = False
    else:
        print(f"  [MISSING] {p}: folder not found")
        all_ok = False

# Verify protocol files are also there
proto_dir = os.path.join(DRIVE_LA, 'ASVspoof2019_LA_cm_protocols')
print(f"\nProtocol folder: {proto_dir}")
if os.path.exists(proto_dir):
    for f in sorted(os.listdir(proto_dir)):
        size_kb = os.path.getsize(os.path.join(proto_dir, f)) / 1024
        print(f"  - {f} ({size_kb:.1f} KB)")
else:
    print("  MISSING")
    all_ok = False

print(f"\nOverall: {'OK' if all_ok else 'PROBLEM - do not delete local copy yet'}")

Verifying files on Drive:
  [OK] train: 25,380 files
  [OK] dev: 24,986 files
  [OK] eval: 71,933 files

Protocol folder: /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_cm_protocols
  - ASVspoof2019.LA.cm.dev.trl.txt (803.1 KB)
  - ASVspoof2019.LA.cm.eval.trl.txt (2302.9 KB)
  - ASVspoof2019.LA.cm.train.trn.txt (820.4 KB)

Overall: OK


In [ ]:
import shutil

LOCAL_DOWNLOAD = '/content/kaggle_download'

if os.path.exists(LOCAL_DOWNLOAD):
    print(f"Removing {LOCAL_DOWNLOAD}...")
    shutil.rmtree(LOCAL_DOWNLOAD)
    print("Done.")
else:
    print("Already removed.")

!df -h /content | tail -1

Removing /content/kaggle_download...
Done.
overlay         236G   51G  186G  22% /


In [ ]:
import os
DRIVE_ROOT = '/content/drive/MyDrive'
if os.path.exists(DRIVE_ROOT):
    print("Drive already mounted.")
else:
    from google.colab import drive
    drive.mount('/content/drive')

DRIVE_LA = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"\nData folder exists: {os.path.exists(DRIVE_LA)}")
print(f"Contents: {sorted(os.listdir(DRIVE_LA))}")


Drive already mounted.

Data folder exists: True
Contents: ['ASVspoof2019_LA_asv_protocols', 'ASVspoof2019_LA_asv_scores', 'ASVspoof2019_LA_cm_protocols', 'ASVspoof2019_LA_dev', 'ASVspoof2019_LA_eval', 'ASVspoof2019_LA_train', 'README.LA.txt']


In [ ]:
import os

PROTO_DIR = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_cm_protocols'
TRAIN_PROTO = os.path.join(PROTO_DIR, 'ASVspoof2019.LA.cm.train.trn.txt')

size_mb = os.path.getsize(TRAIN_PROTO) / 1024
print(f"File: {TRAIN_PROTO}")
print(f"Size: {size_mb:.1f} KB\n")

print("First 10 lines:")
print("-" * 70)
with open(TRAIN_PROTO, 'r') as f:
    for i, line in enumerate(f):
        if i >= 10:
            break
        print(repr(line))
print("-" * 70)

File: /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_cm_protocols/ASVspoof2019.LA.cm.train.trn.txt
Size: 820.4 KB

First 10 lines:
----------------------------------------------------------------------
'LA_0079 LA_T_1138215 - - bonafide\n'
'LA_0079 LA_T_1271820 - - bonafide\n'
'LA_0079 LA_T_1272637 - - bonafide\n'
'LA_0079 LA_T_1276960 - - bonafide\n'
'LA_0079 LA_T_1341447 - - bonafide\n'
'LA_0079 LA_T_1363611 - - bonafide\n'
'LA_0079 LA_T_1596451 - - bonafide\n'
'LA_0079 LA_T_1608170 - - bonafide\n'
'LA_0079 LA_T_1684951 - - bonafide\n'
'LA_0079 LA_T_1699801 - - bonafide\n'
----------------------------------------------------------------------


so bonafide if human and spoof if ai generated

In [ ]:
print("Looking for spoof samples (with attack ID):\n")
print("-" * 70)
count = 0
with open(TRAIN_PROTO, 'r') as f:
    for line in f:
        if 'spoof' in line:
            print(repr(line))
            count += 1
            if count >= 5:
                break
print("-" * 70)

Looking for spoof samples (with attack ID):

----------------------------------------------------------------------
'LA_0079 LA_T_1004644 - A01 spoof\n'
'LA_0079 LA_T_1056709 - A01 spoof\n'
'LA_0079 LA_T_1195221 - A01 spoof\n'
'LA_0079 LA_T_1265032 - A01 spoof\n'
'LA_0079 LA_T_1287124 - A01 spoof\n'
----------------------------------------------------------------------


In [ ]:
from collections import Counter

label_counts = Counter()
attack_counts = Counter()

with open(TRAIN_PROTO, 'r') as f:
    for line in f:
        parts = line.strip().split()
        if len(parts) != 5:
            continue
        speaker, utt_id, _, attack_id, label = parts
        label_counts[label] += 1
        attack_counts[attack_id] += 1

print("Label distribution (train):")
for label, n in label_counts.most_common():
    print(f"  {label}: {n:,}")

print(f"\nTotal: {sum(label_counts.values()):,}")
print(f"Imbalance ratio: 1 bonafide : {label_counts['spoof'] / label_counts['bonafide']:.1f} spoof\n")

print("Attack ID distribution (train):")
for attack, n in sorted(attack_counts.items()):
    label = "bonafide" if attack == "-" else "spoof"
    print(f"  {attack}: {n:,}  ({label})")

Label distribution (train):
  spoof: 22,800
  bonafide: 2,580

Total: 25,380
Imbalance ratio: 1 bonafide : 8.8 spoof

Attack ID distribution (train):
  -: 2,580  (bonafide)
  A01: 3,800  (spoof)
  A02: 3,800  (spoof)
  A03: 3,800  (spoof)
  A04: 3,800  (spoof)
  A05: 3,800  (spoof)
  A06: 3,800  (spoof)


####this demonstrates:
Class imbalance: 1 bona fide for every 8.8 spoofed samples.
If you train a model and just optimize for "raw accuracy," it will quickly learn the laziest possible strategy: always predict "spoof." That gets it ~90% accuracy without learning anything about audio. This is precisely why your proposal called out weighted loss and why anti-spoofing research uses EER instead of accuracy. You can now articulate this from your own data, not from a textbook claim.
Attack distribution: perfectly balanced across A01-A06.
Each of the six training attacks has exactly 3,800 samples. This was a deliberate design choice by the ASVspoof organizers — they wanted no attack type to dominate. Important implication: the model can't shortcut by memorizing one specific attack's artifacts.
A01-A06 are training attacks. A07-A19 will appear in eval.
The eval set contains attacks the model has never seen during training. This is the entire point of the ASVspoof 2019 LA challenge: can your model generalize to unseen attack types? When you eventually report per-attack EER on eval, you'll see different attacks the model handles differently.

In [ ]:
import torchaudio
from IPython.display import Audio, display

TRAIN_FLAC_DIR = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_train/flac'

# A bonafide sample (from the protocol output above)
bonafide_path = f"{TRAIN_FLAC_DIR}/LA_T_1138215.flac"

print("=" * 60)
print("BONAFIDE SAMPLE: LA_T_1138215.flac")
print("=" * 60)
waveform, sr = torchaudio.load(bonafide_path)
print(f"Shape: {waveform.shape}  (channels x samples)")
print(f"Sample rate: {sr} Hz")
print(f"Duration: {waveform.shape[1] / sr:.2f} seconds")
print(f"Min/max amplitude: {waveform.min():.3f} / {waveform.max():.3f}")
display(Audio(waveform.numpy(), rate=sr))

print("\n" + "=" * 60)
print("SPOOF SAMPLE (attack A01): LA_T_1004644.flac")
print("=" * 60)
spoof_path = f"{TRAIN_FLAC_DIR}/LA_T_1004644.flac"
waveform_s, sr_s = torchaudio.load(spoof_path)
print(f"Shape: {waveform_s.shape}")
print(f"Sample rate: {sr_s} Hz")
print(f"Duration: {waveform_s.shape[1] / sr_s:.2f} seconds")
print(f"Min/max amplitude: {waveform_s.min():.3f} / {waveform_s.max():.3f}")
display(Audio(waveform_s.numpy(), rate=sr_s))

BONAFIDE SAMPLE: LA_T_1138215.flac
Shape: torch.Size([1, 55329])  (channels x samples)
Sample rate: 16000 Hz
Duration: 3.46 seconds
Min/max amplitude: -1.000 / 0.974



SPOOF SAMPLE (attack A01): LA_T_1004644.flac
Shape: torch.Size([1, 30769])
Sample rate: 16000 Hz
Duration: 1.92 seconds
Min/max amplitude: -1.000 / 0.611


What you just experienced is the entire point of the field
Modern voice cloning is good enough that human ears are unreliable judges. Researchers have run formal studies on this — when shown bona fide vs spoofed clips, untrained listeners get accuracy in the 55-70% range (barely above the 50% random-guess baseline for binary classification). With training and headphones, it goes up some, but not dramatically.
This isn't a flaw in your hearing. It's the threat. Voice cloning fraud works precisely because humans can't tell.
Why the bona fide sounded "more AI" to you
A few possibilities, all interesting:

Bona fide ASVspoof recordings were studio-recorded. The bona fide samples were collected in clean recording booths with high-quality mics — they sound unnaturally clean, no room reverb, no background noise. Real conversations don't sound like that. So clean audio can register to your brain as "fake-sounding" because it lacks the imperfections we expect.
Modern TTS often adds learned naturalness. Newer synthesis systems (some of A01-A06) were trained on data that included natural speech patterns — small breaths, hesitations, micro-imperfections. They can sound more like everyday speech than a sterile studio recording.
You don't know what the speaker normally sounds like. When we judge "this sounds AI," we're often comparing against an internal expectation of what natural speech sounds like — and that internal model is wrong about studio audio.

Why this matters for your project
This is genuinely useful for your report's introduction and discussion sections. You can write something like:

"Informal listening tests during dataset exploration revealed that spoofed samples in ASVspoof 2019 LA can sound more natural to a casual listener than the corresponding studio-recorded bona fide samples. This observation motivates the need for automated detection systems that can identify subtle synthesis artifacts not perceptible to the human ear."

That's a strong, honest opening framing. It comes from your own engagement with the data, not a textbook claim.
What the model will actually use
Your model — and any anti-spoofing system — doesn't rely on "does this sound natural?" It looks for specific artifacts of the synthesis process:

Phase inconsistencies — synthesizers can produce phase relationships that don't occur in natural speech
High-frequency artifacts — neural vocoders sometimes leave checkerboard patterns or dead bands above ~7 kHz
Spectral envelope anomalies — formant transitions that are slightly too smooth or too sharp
Pitch periodicity — overly regular pitch contours

These are below the threshold of human hearing in many cases. Wav2Vec 2.0 sees the raw waveform and learns to detect them.


In [ ]:
from dataclasses import dataclass
from typing import List
import os

@dataclass
class Utterance:
    """One row from a protocol file."""
    speaker_id: str       # e.g., "LA_0079"
    utterance_id: str     # e.g., "LA_T_1138215"
    attack_id: str        # "-" for bonafide, "A01"-"A19" for spoof
    label: str            # "bonafide" or "spoof"
    label_int: int        # 0 = bonafide, 1 = spoof
    flac_path: str        # full path to the .flac file


def parse_protocol(protocol_path: str, audio_root: str) -> List[Utterance]:
    """
    Parse an ASVspoof 2019 LA cm protocol file.

    Args:
        protocol_path: full path to the .txt protocol file
        audio_root: full path to the folder containing .flac files
                    (e.g., .../ASVspoof2019_LA_train/flac)

    Returns:
        List of Utterance objects.
    """
    utterances = []
    with open(protocol_path, 'r') as f:
        for line_num, line in enumerate(f, start=1):
            parts = line.strip().split()
            if len(parts) != 5:
                # Skip malformed lines
                continue
            speaker_id, utt_id, _unused, attack_id, label = parts

            label_int = 0 if label == 'bonafide' else 1

            flac_path = os.path.join(audio_root, f"{utt_id}.flac")

            utterances.append(Utterance(
                speaker_id=speaker_id,
                utterance_id=utt_id,
                attack_id=attack_id,
                label=label,
                label_int=label_int,
                flac_path=flac_path,
            ))
    return utterances

LA_ROOT = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
train_proto = f"{LA_ROOT}/ASVspoof2019_LA_cm_protocols/ASVspoof2019.LA.cm.train.trn.txt"
train_audio = f"{LA_ROOT}/ASVspoof2019_LA_train/flac"

train_utterances = parse_protocol(train_proto, train_audio)

print(f"Parsed {len(train_utterances):,} utterances from training protocol\n")
print("First 3 entries:")
for u in train_utterances[:3]:
    print(f"  {u}")
print(f"\nFirst spoof entry:")
for u in train_utterances:
    if u.label == 'spoof':
        print(f"  {u}")
        break

Parsed 25,380 utterances from training protocol

First 3 entries:
  Utterance(speaker_id='LA_0079', utterance_id='LA_T_1138215', attack_id='-', label='bonafide', label_int=0, flac_path='/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_train/flac/LA_T_1138215.flac')
  Utterance(speaker_id='LA_0079', utterance_id='LA_T_1271820', attack_id='-', label='bonafide', label_int=0, flac_path='/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_train/flac/LA_T_1271820.flac')
  Utterance(speaker_id='LA_0079', utterance_id='LA_T_1272637', attack_id='-', label='bonafide', label_int=0, flac_path='/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_train/flac/LA_T_1272637.flac')

First spoof entry:
  Utterance(speaker_id='LA_0079', utterance_id='LA_T_1004644', attack_id='A01', label='spoof', label_int=1, flac_path='/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA/ASVspoof2019_LA_train/flac/LA_T_100464

LA_0079 appears in both bonafide and spoof entries. That's intentional in ASVspoof — the same speaker has both real recordings AND spoofed versions (synthesizers were given samples of LA_0079's voice and asked to mimic them). The model needs to learn "this is fake regardless of who it claims to be

In [ ]:
def parse_all_partitions(la_root: str):
    """Parse train, dev, and eval protocol files into a dict."""
    proto_dir = f"{la_root}/ASVspoof2019_LA_cm_protocols"

    partitions = {
        'train': {
            'proto': f"{proto_dir}/ASVspoof2019.LA.cm.train.trn.txt",
            'audio': f"{la_root}/ASVspoof2019_LA_train/flac",
        },
        'dev': {
            'proto': f"{proto_dir}/ASVspoof2019.LA.cm.dev.trl.txt",
            'audio': f"{la_root}/ASVspoof2019_LA_dev/flac",
        },
        'eval': {
            'proto': f"{proto_dir}/ASVspoof2019.LA.cm.eval.trl.txt",
            'audio': f"{la_root}/ASVspoof2019_LA_eval/flac",
        },
    }

    result = {}
    for name, paths in partitions.items():
        utterances = parse_protocol(paths['proto'], paths['audio'])
        result[name] = utterances
    return result


# Parse everything
splits = parse_all_partitions(LA_ROOT)

# Summary table
print(f"{'Partition':<10} {'Total':>10} {'Bonafide':>10} {'Spoof':>10} {'Ratio':>10}")
print("-" * 56)
for name, utts in splits.items():
    n_total = len(utts)
    n_bonafide = sum(1 for u in utts if u.label == 'bonafide')
    n_spoof = n_total - n_bonafide
    ratio = n_spoof / n_bonafide if n_bonafide > 0 else 0
    print(f"{name:<10} {n_total:>10,} {n_bonafide:>10,} {n_spoof:>10,} {ratio:>9.1f}x")

# Show attack ID distribution per partition
from collections import Counter
print("\nAttack IDs per partition:")
for name, utts in splits.items():
    attack_counts = Counter(u.attack_id for u in utts if u.attack_id != '-')
    attacks = sorted(attack_counts.keys())
    print(f"  {name}: {attacks}")

Partition       Total   Bonafide      Spoof      Ratio
--------------------------------------------------------
train          25,380      2,580     22,800       8.8x
dev            24,844      2,548     22,296       8.8x
eval           71,237      7,355     63,882       8.7x

Attack IDs per partition:
  train: ['A01', 'A02', 'A03', 'A04', 'A05', 'A06']
  dev: ['A01', 'A02', 'A03', 'A04', 'A05', 'A06']
  eval: ['A07', 'A08', 'A09', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A16', 'A17', 'A18', 'A19']


In [ ]:
import os

# Verify the repo is cloned in Colab
REPO_DIR = '/content/deepfake-audio-detection'
if not os.path.exists(REPO_DIR):
    print("Repo not found. Cloning...")
    !git clone https://github.com/Saracasm/deepfake-audio-detection.git {REPO_DIR}
else:
    print(f"Repo found at {REPO_DIR}")
    !cd {REPO_DIR} && git pull --quiet

# Make sure src/data/ exists with __init__.py
os.makedirs(f'{REPO_DIR}/src/data', exist_ok=True)
for init_path in [f'{REPO_DIR}/src/__init__.py', f'{REPO_DIR}/src/data/__init__.py']:
    if not os.path.exists(init_path):
        open(init_path, 'w').close()

print(f"Repo ready at {REPO_DIR}")

Repo not found. Cloning...
Cloning into '/content/deepfake-audio-detection'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (36/36), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 36 (delta 12), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (36/36), 18.88 KiB | 2.36 MiB/s, done.
Resolving deltas: 100% (12/12), done.
Repo ready at /content/deepfake-audio-detection


In [ ]:
PROTOCOLS_PY = '''"""
ASVspoof 2019 LA protocol parser.

Reads the official .txt protocol files and yields structured Utterance objects
that downstream code (datasets, evaluation) can use.

Protocol file format (5 space-separated columns):
    speaker_id  utterance_id  -  attack_id  label

    speaker_id   : anonymized speaker (e.g., "LA_0079")
    utterance_id : filename without extension (e.g., "LA_T_1138215")
    column 3     : unused, always "-"
    attack_id    : "-" for bonafide, "A01"-"A19" for spoof samples
    label        : "bonafide" or "spoof"
"""

from dataclasses import dataclass
from typing import List, Dict
import os


@dataclass
class Utterance:
    """One row from an ASVspoof 2019 LA protocol file."""
    speaker_id: str
    utterance_id: str
    attack_id: str        # "-" for bonafide, "A01"-"A19" for spoof
    label: str            # "bonafide" or "spoof"
    label_int: int        # 0 = bonafide, 1 = spoof
    flac_path: str        # absolute path to the .flac file


def parse_protocol(protocol_path: str, audio_root: str) -> List[Utterance]:
    """Parse one ASVspoof 2019 LA cm protocol file.

    Args:
        protocol_path: full path to the .txt protocol file.
        audio_root: full path to the folder containing the .flac files.

    Returns:
        List of Utterance objects, one per valid line.
    """
    utterances: List[Utterance] = []
    with open(protocol_path, "r") as f:
        for line in f:
            parts = line.strip().split()
            if len(parts) != 5:
                continue
            speaker_id, utt_id, _unused, attack_id, label = parts
            label_int = 0 if label == "bonafide" else 1
            flac_path = os.path.join(audio_root, f"{utt_id}.flac")
            utterances.append(Utterance(
                speaker_id=speaker_id,
                utterance_id=utt_id,
                attack_id=attack_id,
                label=label,
                label_int=label_int,
                flac_path=flac_path,
            ))
    return utterances


def parse_all_partitions(la_root: str) -> Dict[str, List[Utterance]]:
    """Parse train, dev, and eval protocols at once.

    Args:
        la_root: path to the LA folder, e.g.
                 ".../asvspoof_2019/LA"

    Returns:
        Dict with keys "train", "dev", "eval" mapping to lists of Utterances.
    """
    proto_dir = os.path.join(la_root, "ASVspoof2019_LA_cm_protocols")
    partitions = {
        "train": (
            os.path.join(proto_dir, "ASVspoof2019.LA.cm.train.trn.txt"),
            os.path.join(la_root, "ASVspoof2019_LA_train", "flac"),
        ),
        "dev": (
            os.path.join(proto_dir, "ASVspoof2019.LA.cm.dev.trl.txt"),
            os.path.join(la_root, "ASVspoof2019_LA_dev", "flac"),
        ),
        "eval": (
            os.path.join(proto_dir, "ASVspoof2019.LA.cm.eval.trl.txt"),
            os.path.join(la_root, "ASVspoof2019_LA_eval", "flac"),
        ),
    }
    return {
        name: parse_protocol(proto, audio)
        for name, (proto, audio) in partitions.items()
    }


def class_counts(utterances: List[Utterance]) -> Dict[str, int]:
    """Return {'bonafide': N, 'spoof': M} counts."""
    counts = {"bonafide": 0, "spoof": 0}
    for u in utterances:
        counts[u.label] += 1
    return counts
'''

PATH = '/content/deepfake-audio-detection/src/data/protocols.py'
with open(PATH, 'w') as f:
    f.write(PROTOCOLS_PY)

print(f"Wrote {PATH}")
print(f"Size: {os.path.getsize(PATH)} bytes")
print(f"Lines: {len(PROTOCOLS_PY.splitlines())}")

Wrote /content/deepfake-audio-detection/src/data/protocols.py
Size: 3284 bytes
Lines: 99


In [ ]:
import sys
import importlib

# Add repo to Python path so we can import from src/
sys.path.insert(0, '/content/deepfake-audio-detection')

# Force reload in case we re-run
if 'src.data.protocols' in sys.modules:
    importlib.reload(sys.modules['src.data.protocols'])

from src.data.protocols import parse_all_partitions, class_counts

# Re-parse using the module
splits = parse_all_partitions(LA_ROOT)

print("Re-parsed via module:")
for name, utts in splits.items():
    counts = class_counts(utts)
    print(f"  {name}: {len(utts):,} total ({counts['bonafide']:,} bonafide, {counts['spoof']:,} spoof)")

print("\nModule works.")

Re-parsed via module:
  train: 25,380 total (2,580 bonafide, 22,800 spoof)
  dev: 24,844 total (2,548 bonafide, 22,296 spoof)
  eval: 71,237 total (7,355 bonafide, 63,882 spoof)

Module works.


In [ ]:
!cd /content/deepfake-audio-detection && git config user.email "95262824+Saracasm@users.noreply.github.com" && git config user.name "Sara Iqbal"
print("Git identity configured.")

Git identity configured.


In [ ]:
%cd /content/deepfake-audio-detection
!git add src/data/protocols.py src/__init__.py src/data/__init__.py
!git status

/content/deepfake-audio-detection
On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   src/__init__.py
	new file:   src/data/__init__.py
	new file:   src/data/protocols.py



In [ ]:
!git commit -m "Phase 2: add ASVspoof 2019 LA protocol parser"

[main a14fd25] Phase 2: add ASVspoof 2019 LA protocol parser
 3 files changed, 99 insertions(+)
 create mode 100644 src/__init__.py
 create mode 100644 src/data/__init__.py
 create mode 100644 src/data/protocols.py


In [ ]:
GITHUB_TOKEN = "REDACTED_TOKEN"
GITHUB_USER = "Saracasm"
REPO = "deepfake-audio-detection"

push_url = f"https://{GITHUB_USER}:{GITHUB_TOKEN}@github.com/{GITHUB_USER}/{REPO}.git"
!cd /content/deepfake-audio-detection && git push {push_url} main

print("\nDone.")

Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 2 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (6/6), 1.58 KiB | 1.58 MiB/s, done.
Total 6 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/Saracasm/deepfake-audio-detection.git
   d9f77a5..a14fd25  main -> main

Done.


In [ ]:
GITHUB_TOKEN = "REDACTED_TOKEN"

In [ ]:
import torchaudio
import numpy as np
from tqdm import tqdm
import random

# Sample 500 random training clips and measure durations
random.seed(42)
sample = random.sample(splits['train'], 500)

durations = []
for u in tqdm(sample, desc="Measuring durations"):
    # Load audio (this works across all torchaudio versions)
    waveform, sr = torchaudio.load(u.flac_path)
    duration_sec = waveform.shape[1] / sr
    durations.append(duration_sec)

durations = np.array(durations)

print(f"\nDuration statistics over {len(durations)} random training clips:")
print(f"  Min:         {durations.min():.2f} sec")
print(f"  Max:         {durations.max():.2f} sec")
print(f"  Mean:        {durations.mean():.2f} sec")
print(f"  Median:      {np.median(durations):.2f} sec")
print(f"  Std dev:     {durations.std():.2f} sec")
print(f"\nPercentiles:")
print(f"  10%:  {np.percentile(durations, 10):.2f} sec")
print(f"  25%:  {np.percentile(durations, 25):.2f} sec")
print(f"  50%:  {np.percentile(durations, 50):.2f} sec")
print(f"  75%:  {np.percentile(durations, 75):.2f} sec")
print(f"  90%:  {np.percentile(durations, 90):.2f} sec")
print(f"\nClips shorter than 4 sec: {(durations < 4).sum()} / {len(durations)} ({100*(durations < 4).mean():.1f}%)")
print(f"Clips longer than 4 sec:  {(durations > 4).sum()} / {len(durations)} ({100*(durations > 4).mean():.1f}%)")

Measuring durations: 100%|██████████| 500/500 [00:01<00:00, 453.95it/s]


Duration statistics over 500 random training clips:
  Min:         0.82 sec
  Max:         10.28 sec
  Mean:        3.52 sec
  Median:      3.24 sec
  Std dev:     1.44 sec

Percentiles:
  10%:  1.93 sec
  25%:  2.51 sec
  50%:  3.24 sec
  75%:  4.32 sec
  90%:  5.48 sec

Clips shorter than 4 sec: 346 / 500 (69.2%)
Clips longer than 4 sec:  154 / 500 (30.8%)


In [ ]:
PREPROCESSING_PY = '''"""
Audio preprocessing for ASVspoof 2019 LA.

Given a .flac path, produces a list of fixed-length 4-second segments at 16kHz
mono, ready to feed into Wav2Vec 2.0.

Pipeline:
    load .flac -> ensure mono -> ensure 16kHz -> window with 50% overlap
    (short clips are zero-padded to one full window)
"""

from typing import List
import torch
import torchaudio
import torchaudio.functional as F


SAMPLE_RATE = 16000
WINDOW_SECONDS = 4.0
OVERLAP_RATIO = 0.5

WINDOW_SAMPLES = int(SAMPLE_RATE * WINDOW_SECONDS)              # 64000
HOP_SAMPLES = int(WINDOW_SAMPLES * (1.0 - OVERLAP_RATIO))       # 32000


def load_audio(path: str, target_sr: int = SAMPLE_RATE) -> torch.Tensor:
    """Load a .flac file and return a 1-D mono waveform at target_sr.

    Returns:
        Tensor of shape (num_samples,) at target_sr.
    """
    waveform, sr = torchaudio.load(path)  # shape: (channels, num_samples)

    # Ensure mono: average if multi-channel
    if waveform.shape[0] > 1:
        waveform = waveform.mean(dim=0, keepdim=True)

    # Ensure target sample rate
    if sr != target_sr:
        waveform = F.resample(waveform, orig_freq=sr, new_freq=target_sr)

    return waveform.squeeze(0)  # shape: (num_samples,)


def segment_waveform(
    waveform: torch.Tensor,
    window_samples: int = WINDOW_SAMPLES,
    hop_samples: int = HOP_SAMPLES,
) -> List[torch.Tensor]:
    """Split a 1-D waveform into fixed-length windows with overlap.

    Short clips (< window_samples) are zero-padded to one full window.
    Long clips are windowed with the given hop size; the last window
    may be partially padded if it does not fit exactly.

    Returns:
        List of tensors, each of shape (window_samples,).
    """
    n = waveform.shape[0]

    # Short clip: zero-pad to one full window
    if n <= window_samples:
        padded = torch.zeros(window_samples, dtype=waveform.dtype)
        padded[:n] = waveform
        return [padded]

    # Long clip: walk through with hop, padding the final window if needed
    windows = []
    start = 0
    while start < n:
        end = start + window_samples
        if end <= n:
            windows.append(waveform[start:end])
        else:
            # Last window — pad with zeros up to window_samples
            tail = waveform[start:]
            padded = torch.zeros(window_samples, dtype=waveform.dtype)
            padded[:tail.shape[0]] = tail
            windows.append(padded)
            break  # done after the last partial window
        start += hop_samples

    return windows


def preprocess(path: str) -> List[torch.Tensor]:
    """Full pipeline: load + window. Convenience wrapper."""
    waveform = load_audio(path)
    return segment_waveform(waveform)
'''

PATH = '/content/deepfake-audio-detection/src/data/preprocessing.py'
with open(PATH, 'w') as f:
    f.write(PREPROCESSING_PY)

print(f"Wrote {PATH}")
print(f"Size: {len(PREPROCESSING_PY)} bytes")

Wrote /content/deepfake-audio-detection/src/data/preprocessing.py
Size: 2720 bytes


In [ ]:
import torch
import importlib
import src.data.preprocessing
importlib.reload(src.data.preprocessing)
from src.data.preprocessing import load_audio, segment_waveform, preprocess, WINDOW_SAMPLES, HOP_SAMPLES

print(f"Window: {WINDOW_SAMPLES} samples ({WINDOW_SAMPLES/16000:.1f} sec)")
print(f"Hop:    {HOP_SAMPLES} samples ({HOP_SAMPLES/16000:.1f} sec)\n")

# --- Pick three test clips of different lengths ---
# We already have `splits` from earlier, with `train` utterances loaded
# We'll pick a known short clip, a medium clip, and find a long one
print("Finding test clips of varying length...")

short_clip = None
medium_clip = None
long_clip = None

# Sample some clips and bucket them
for u in splits['train'][:200]:
    w = load_audio(u.flac_path)
    duration = w.shape[0] / 16000
    if duration < 2.0 and short_clip is None:
        short_clip = (u, w, duration)
    elif 3.5 < duration < 4.5 and medium_clip is None:
        medium_clip = (u, w, duration)
    elif duration > 7.0 and long_clip is None:
        long_clip = (u, w, duration)
    if all([short_clip, medium_clip, long_clip]):
        break

# --- Test each ---
print("\n" + "=" * 70)
print("TEST 1: Short clip (should produce 1 window with padding)")
print("=" * 70)
if short_clip:
    u, w, duration = short_clip
    print(f"  File: {u.utterance_id}")
    print(f"  Original duration: {duration:.2f} sec ({w.shape[0]} samples)")
    windows = segment_waveform(w)
    print(f"  Windows produced: {len(windows)}")
    print(f"  Each window shape: {windows[0].shape}")
    # Check: real audio at start, zeros at end
    n_real = w.shape[0]
    print(f"  First {n_real} samples match original: {torch.allclose(windows[0][:n_real], w)}")
    print(f"  Padding zeros at end: {(windows[0][n_real:] == 0).all().item()}")
else:
    print("  No clip under 2 sec found in first 200 utterances")

print("\n" + "=" * 70)
print("TEST 2: Medium clip (~4 sec, should produce 1 window)")
print("=" * 70)
if medium_clip:
    u, w, duration = medium_clip
    print(f"  File: {u.utterance_id}")
    print(f"  Original duration: {duration:.2f} sec ({w.shape[0]} samples)")
    windows = segment_waveform(w)
    print(f"  Windows produced: {len(windows)}")
    print(f"  Each window shape: {windows[0].shape}")

print("\n" + "=" * 70)
print("TEST 3: Long clip (>7 sec, should produce multiple windows)")
print("=" * 70)
if long_clip:
    u, w, duration = long_clip
    print(f"  File: {u.utterance_id}")
    print(f"  Original duration: {duration:.2f} sec ({w.shape[0]} samples)")
    windows = segment_waveform(w)
    print(f"  Windows produced: {len(windows)}")
    print(f"  Each window shape: {windows[0].shape}")
    # Predict expected: ceil((n - window) / hop) + 1
    expected = max(1, ((w.shape[0] - WINDOW_SAMPLES + HOP_SAMPLES - 1) // HOP_SAMPLES) + 1)
    print(f"  Expected windows: {expected}")
else:
    print("  No clip over 7 sec found in first 200 utterances")

print("\n" + "=" * 70)
print("Preprocessing tests complete.")
print("=" * 70)

Window: 64000 samples (4.0 sec)
Hop:    32000 samples (2.0 sec)

Finding test clips of varying length...

TEST 1: Short clip (should produce 1 window with padding)
  File: LA_T_9701712
  Original duration: 2.00 sec (31966 samples)
  Windows produced: 1
  Each window shape: torch.Size([64000])
  First 31966 samples match original: True
  Padding zeros at end: True

TEST 2: Medium clip (~4 sec, should produce 1 window)
  File: LA_T_1271820
  Original duration: 4.40 sec (70323 samples)
  Windows produced: 2
  Each window shape: torch.Size([64000])

TEST 3: Long clip (>7 sec, should produce multiple windows)
  No clip over 7 sec found in first 200 utterances

Preprocessing tests complete.


In [ ]:
# Search a wider sample to find a clip > 7 seconds
print("Searching for a long clip (>7 sec) in first 5,000 utterances...")
long_clip = None
for u in splits['train'][:5000]:
    w = load_audio(u.flac_path)
    duration = w.shape[0] / 16000
    if duration > 7.0:
        long_clip = (u, w, duration)
        break

if long_clip:
    u, w, duration = long_clip
    print(f"\nFound: {u.utterance_id}")
    print(f"Original duration: {duration:.2f} sec ({w.shape[0]} samples)")
    windows = segment_waveform(w)
    print(f"Windows produced: {len(windows)}")

    # Show how each window aligns with the original audio
    print(f"\nWindow alignment:")
    for i, win in enumerate(windows):
        start_sample = i * HOP_SAMPLES
        end_sample = start_sample + WINDOW_SAMPLES
        real_samples = min(WINDOW_SAMPLES, w.shape[0] - start_sample)
        padded_samples = WINDOW_SAMPLES - real_samples
        start_sec = start_sample / 16000
        end_sec = end_sample / 16000
        print(f"  Window {i}: covers {start_sec:.2f}-{end_sec:.2f} sec  "
              f"({real_samples} real + {padded_samples} padded samples)")
else:
    print("Still no clip > 7 sec. The training set may not contain very long clips.")
    print("Skipping test 3 — preprocessing logic is verified by tests 1 and 2 anyway.")

Searching for a long clip (>7 sec) in first 5,000 utterances...

Found: LA_T_1963829
Original duration: 7.38 sec (118059 samples)
Windows produced: 3

Window alignment:
  Window 0: covers 0.00-4.00 sec  (64000 real + 0 padded samples)
  Window 1: covers 2.00-6.00 sec  (64000 real + 0 padded samples)
  Window 2: covers 4.00-8.00 sec  (54059 real + 9941 padded samples)


In [ ]:
import shutil
import os

# Find the notebook in Drive
drive_notebook = '/content/drive/MyDrive/Colab Notebooks/01_data_acquisition.ipynb'
repo_notebook = '/content/deepfake-audio-detection/notebooks/01_data_acquisition.ipynb'

# Make sure target folder exists
os.makedirs(os.path.dirname(repo_notebook), exist_ok=True)

# Verify source exists (notebook may have a slightly different name if you renamed it)
if not os.path.exists(drive_notebook):
    print(f"NOT FOUND: {drive_notebook}")
    print("\nLet me list what's actually in Colab Notebooks/ so we can find it:")
    drive_dir = '/content/drive/MyDrive/Colab Notebooks'
    if os.path.exists(drive_dir):
        for f in sorted(os.listdir(drive_dir)):
            print(f"  - {f}")
else:
    shutil.copy(drive_notebook, repo_notebook)
    size_kb = os.path.getsize(repo_notebook) / 1024
    print(f"Copied notebook to repo: {repo_notebook}")
    print(f"Size: {size_kb:.1f} KB")

NOT FOUND: /content/drive/MyDrive/Colab Notebooks/01_data_acquisition.ipynb

Let me list what's actually in Colab Notebooks/ so we can find it:


In [ ]:
import os

# Search the entire Drive for any .ipynb files
print("Searching for .ipynb files in Drive...")
print("(may take 30 seconds on a big Drive)\n")

found = []
for root, dirs, files in os.walk('/content/drive/MyDrive'):
    # Skip the dataset folder (lots of files, no notebooks)
    if 'deepfake_audio' in root and 'data' in root:
        dirs.clear()  # don't descend
        continue
    for f in files:
        if f.endswith('.ipynb'):
            full = os.path.join(root, f)
            size_kb = os.path.getsize(full) / 1024
            found.append((full, size_kb))

if found:
    print(f"Found {len(found)} notebook(s):\n")
    for path, size in found:
        print(f"  [{size:>7.1f} KB]  {path}")
else:
    print("No .ipynb files found in Drive.")
    print("\nThis means the notebook is only in Colab's session storage,")
    print("not actually saved to Drive. We need to fix that.")

Searching for .ipynb files in Drive...
(may take 30 seconds on a big Drive)

No .ipynb files found in Drive.

This means the notebook is only in Colab's session storage,
not actually saved to Drive. We need to fix that.


In [ ]:
import os

print("=== Drive (persists across disconnects) ===")
drive_la = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"  Dataset on Drive: {os.path.exists(drive_la)}")

print("\n=== Colab local disk (wiped on disconnect) ===")
repo_dir = '/content/deepfake-audio-detection'
print(f"  Repo cloned:       {os.path.exists(repo_dir)}")
preprocessing_py = f'{repo_dir}/src/data/preprocessing.py'
dataset_py = f'{repo_dir}/src/data/dataset.py'
print(f"  preprocessing.py:  {os.path.exists(preprocessing_py)}")
print(f"  dataset.py:        {os.path.exists(dataset_py)}")

print("\n=== Python session memory ===")
try:
    print(f"  splits variable:   exists, train has {len(splits['train'])} utterances")
except NameError:
    print(f"  splits variable:   GONE (need to re-parse)")

try:
    print(f"  LA_ROOT defined:   {LA_ROOT}")
except NameError:
    print(f"  LA_ROOT defined:   GONE")

=== Drive (persists across disconnects) ===
  Dataset on Drive: False

=== Colab local disk (wiped on disconnect) ===
  Repo cloned:       False
  preprocessing.py:  False
  dataset.py:        False

=== Python session memory ===
  splits variable:   GONE (need to re-parse)
  LA_ROOT defined:   GONE


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import os
DRIVE_LA = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"Dataset on Drive: {os.path.exists(DRIVE_LA)}")
print(f"Contents: {sorted(os.listdir(DRIVE_LA))[:3] if os.path.exists(DRIVE_LA) else 'N/A'}")

ValueError: mount failed

In [ ]:
from google.colab import drive
drive.flush_and_unmount()
drive.mount('/content/drive', force_remount=True)

Drive not mounted, so nothing to flush and unmount.
Mounted at /content/drive


In [ ]:
import os
DRIVE_LA = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"Dataset on Drive: {os.path.exists(DRIVE_LA)}")
if os.path.exists(DRIVE_LA):
    print(f"Contents: {sorted(os.listdir(DRIVE_LA))}")

Dataset on Drive: False


In [ ]:
import os

# Check what's actually at the root of MyDrive
print("=== Top-level of MyDrive ===")
mydrive = '/content/drive/MyDrive'
if os.path.exists(mydrive):
    for item in sorted(os.listdir(mydrive))[:30]:
        full = os.path.join(mydrive, item)
        kind = "FOLDER" if os.path.isdir(full) else "file"
        print(f"  [{kind}]  {item}")
else:
    print("  MyDrive itself doesn't exist?")

# Specifically look for our project folder
print("\n=== Looking for deepfake_audio folder ===")
target = '/content/drive/MyDrive/deepfake_audio'
print(f"  Exists: {os.path.exists(target)}")
if os.path.exists(target):
    print(f"  Contents:")
    for item in sorted(os.listdir(target)):
        print(f"    - {item}")

=== Top-level of MyDrive ===
  [file]  0802-python-tutorial.pdf
  [file]  1000026170.jpg
  [file]  1000026171.jpg
  [file]  1000026172.jpg
  [file]  1000026173.jpg
  [file]  1000026174.jpg
  [file]  1000026176.jpg
  [file]  1000026178.jpg
  [file]  7.png
  [file]  A Complete Guide to Programming C++.pdf
  [file]  ADDICTION.pdf
  [file]  Abnormal Psychology.pdf
  [file]  Beginning Programming with Python For Dummies Mueller, John Paul [SRG].pdf
  [file]  Beiratkozási Lap Nyomtatvány - Többnyelvű new.pdf
  [FOLDER]  Colab Notebooks
  [FOLDER]  Colab-Notebooks
  [file]  DLP-Project.gprj
  [file]  DS LAB HOME TASKS.docx
  [file]  DSC_0277.JPG
  [file]  DSC_0278.JPG
  [file]  DSC_0279.JPG
  [file]  DSC_0409.JPG
  [file]  LECTURE 3.pptx
  [file]  LECTURE PERSONALITY (1).pdf
  [file]  Leadership.pdf
  [file]  Lecture 2 RESEARCH METHODS.pdf
  [FOLDER]  Mortgage Docs
  [file]  Name Sara IqbalStudent ID 23K-0669Class CS-5G.pdf
  [FOLDER]  OOP concept texts
  [file]  Resume-Sara-Iqbal.pdf

==

In [ ]:
import os

print("=== Drilling into deepfake_audio/data/ ===")
data = '/content/drive/MyDrive/deepfake_audio/data'
for item in sorted(os.listdir(data)):
    full = os.path.join(data, item)
    print(f"  {item}/")
    if os.path.isdir(full):
        for sub in sorted(os.listdir(full))[:5]:
            print(f"    - {sub}")

print("\n=== Verifying ASVspoof 2019 LA structure ===")
LA_ROOT = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"  Exists: {os.path.exists(LA_ROOT)}")
if os.path.exists(LA_ROOT):
    contents = sorted(os.listdir(LA_ROOT))
    print(f"  Contents: {contents}")


=== Drilling into deepfake_audio/data/ ===
  processed/
  raw/
    - asvspoof_2019
    - asvspoof_2021
    - wavefake

=== Verifying ASVspoof 2019 LA structure ===
  Exists: False


In [ ]:
import os

asvspoof_dir = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019'
print(f"Path: {asvspoof_dir}")
print(f"Exists: {os.path.exists(asvspoof_dir)}\n")

if os.path.exists(asvspoof_dir):
    contents = sorted(os.listdir(asvspoof_dir))
    print(f"Found {len(contents)} item(s):")
    for item in contents:
        full = os.path.join(asvspoof_dir, item)
        kind = "FOLDER" if os.path.isdir(full) else "file"
        try:
            size = os.path.getsize(full) if not os.path.isdir(full) else sum(
                os.path.getsize(os.path.join(r, f))
                for r, d, fs in os.walk(full)
                for f in fs
            )
            size_str = f"{size/1e9:.2f} GB" if size > 1e9 else f"{size/1e6:.2f} MB" if size > 1e6 else f"{size:,} B"
        except:
            size_str = "?"
        print(f"  [{kind}]  {item}  ({size_str})")

Path: /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019
Exists: True

Found 0 item(s):


In [ ]:
import os

print("=== Full state of deepfake_audio/ ===\n")
root = '/content/drive/MyDrive/deepfake_audio'
for r, dirs, files in os.walk(root):
    indent = '  ' * (r.replace(root, '').count('/'))
    print(f"{indent}{os.path.basename(r) or 'deepfake_audio'}/")
    for f in files:
        size_kb = os.path.getsize(os.path.join(r, f)) / 1024
        print(f"{indent}  - {f} ({size_kb:.1f} KB)")

=== Full state of deepfake_audio/ ===

deepfake_audio/
  data/
    raw/
      asvspoof_2019/
      asvspoof_2021/
      wavefake/
    processed/
  checkpoints/
  logs/


In [ ]:
import os

REPO_DIR = '/content/deepfake-audio-detection'
if os.path.exists(REPO_DIR):
    import shutil
    shutil.rmtree(REPO_DIR)

!git clone https://github.com/Saracasm/deepfake-audio-detection.git {REPO_DIR}
print("\nRepo re-cloned.")

Cloning into '/content/deepfake-audio-detection'...
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (35/35), done.
remote: Total 42 (delta 13), reused 7 (delta 1), pack-reused 0 (from 0)
Receiving objects: 100% (42/42), 20.42 KiB | 20.42 MiB/s, done.
Resolving deltas: 100% (13/13), done.

Repo re-cloned.


In [ ]:
PREPROCESSING_PY = '''"""
Audio preprocessing for ASVspoof 2019 LA.

Given a .flac path, produces a list of fixed-length 4-second segments at 16kHz
mono, ready to feed into Wav2Vec 2.0.

Pipeline:
    load .flac -> ensure mono -> ensure 16kHz -> window with 50% overlap
    (short clips are zero-padded to one full window)
"""

from typing import List
import torch
import torchaudio
import torchaudio.functional as F


SAMPLE_RATE = 16000
WINDOW_SECONDS = 4.0
OVERLAP_RATIO = 0.5

WINDOW_SAMPLES = int(SAMPLE_RATE * WINDOW_SECONDS)              # 64000
HOP_SAMPLES = int(WINDOW_SAMPLES * (1.0 - OVERLAP_RATIO))       # 32000


def load_audio(path: str, target_sr: int = SAMPLE_RATE) -> torch.Tensor:
    """Load a .flac file and return a 1-D mono waveform at target_sr."""
    waveform, sr = torchaudio.load(path)
    if waveform.shape[0] > 1:
        waveform = waveform.mean(dim=0, keepdim=True)
    if sr != target_sr:
        waveform = F.resample(waveform, orig_freq=sr, new_freq=target_sr)
    return waveform.squeeze(0)


def segment_waveform(
    waveform: torch.Tensor,
    window_samples: int = WINDOW_SAMPLES,
    hop_samples: int = HOP_SAMPLES,
) -> List[torch.Tensor]:
    """Split a 1-D waveform into fixed-length windows with overlap."""
    n = waveform.shape[0]
    if n <= window_samples:
        padded = torch.zeros(window_samples, dtype=waveform.dtype)
        padded[:n] = waveform
        return [padded]

    windows = []
    start = 0
    while start < n:
        end = start + window_samples
        if end <= n:
            windows.append(waveform[start:end])
        else:
            tail = waveform[start:]
            padded = torch.zeros(window_samples, dtype=waveform.dtype)
            padded[:tail.shape[0]] = tail
            windows.append(padded)
            break
        start += hop_samples
    return windows


def preprocess(path: str) -> List[torch.Tensor]:
    """Full pipeline: load + window."""
    waveform = load_audio(path)
    return segment_waveform(waveform)
'''

PATH = '/content/deepfake-audio-detection/src/data/preprocessing.py'
os.makedirs(os.path.dirname(PATH), exist_ok=True)
with open(PATH, 'w') as f:
    f.write(PREPROCESSING_PY)
print(f"Wrote {PATH} ({len(PREPROCESSING_PY)} bytes)")

Wrote /content/deepfake-audio-detection/src/data/preprocessing.py (1990 bytes)


In [ ]:
DATASET_PY = '''"""
PyTorch Dataset for ASVspoof 2019 LA.

One Dataset entry = one 4-second audio window + its binary label.
Long utterances expand into multiple entries (one per window).
"""

from typing import List, Tuple
import torch
from torch.utils.data import Dataset

from src.data.protocols import Utterance
from src.data.preprocessing import (
    load_audio,
    segment_waveform,
    WINDOW_SAMPLES,
    HOP_SAMPLES,
)


def _count_windows(duration_samples: int) -> int:
    if duration_samples <= WINDOW_SAMPLES:
        return 1
    n = (duration_samples - WINDOW_SAMPLES + HOP_SAMPLES - 1) // HOP_SAMPLES + 1
    return max(1, n)


class ASVspoofDataset(Dataset):
    """One sample = one 4-sec window + its label."""

    def __init__(
        self,
        utterances: List[Utterance],
        durations_samples: List[int] = None,
    ):
        self.utterances = utterances
        self.index: List[Tuple[int, int]] = []
        if durations_samples is None:
            for i in range(len(utterances)):
                self.index.append((i, 0))
        else:
            assert len(durations_samples) == len(utterances)
            for i, d in enumerate(durations_samples):
                n_windows = _count_windows(d)
                for w in range(n_windows):
                    self.index.append((i, w))

    def __len__(self) -> int:
        return len(self.index)

    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, int, str]:
        utt_idx, window_idx = self.index[idx]
        utt = self.utterances[utt_idx]
        waveform = load_audio(utt.flac_path)
        windows = segment_waveform(waveform)
        window_idx = min(window_idx, len(windows) - 1)
        return windows[window_idx], utt.label_int, utt.utterance_id
'''

PATH = '/content/deepfake-audio-detection/src/data/dataset.py'
with open(PATH, 'w') as f:
    f.write(DATASET_PY)
print(f"Wrote {PATH} ({len(DATASET_PY)} bytes)")

Wrote /content/deepfake-audio-detection/src/data/dataset.py (1741 bytes)


In [ ]:
import os
files = [
    '/content/deepfake-audio-detection/src/data/protocols.py',
    '/content/deepfake-audio-detection/src/data/preprocessing.py',
    '/content/deepfake-audio-detection/src/data/dataset.py',
]
for f in files:
    exists = os.path.exists(f)
    size = os.path.getsize(f) if exists else 0
    print(f"  {'OK' if exists else 'MISSING'}  {f}  ({size} bytes)")

  OK  /content/deepfake-audio-detection/src/data/protocols.py  (3284 bytes)
  OK  /content/deepfake-audio-detection/src/data/preprocessing.py  (1990 bytes)
  OK  /content/deepfake-audio-detection/src/data/dataset.py  (1741 bytes)


In [ ]:
from google.colab import files
print("Click 'Choose Files' and select kaggle.json from your Downloads folder")
uploaded = files.upload()

Click 'Choose Files' and select kaggle.json from your Downloads folder


Saving kaggle.json to kaggle.json


In [ ]:
import os
import shutil

os.makedirs('/root/.kaggle', exist_ok=True)
shutil.move('kaggle.json', '/root/.kaggle/kaggle.json')
os.chmod('/root/.kaggle/kaggle.json', 0o600)

!pip install -q kaggle
print("Kaggle CLI ready.")

Kaggle CLI ready.


In [ ]:
import os, time

DOWNLOAD_DIR = '/content/kaggle_download'
os.makedirs(DOWNLOAD_DIR, exist_ok=True)

print("Downloading ASVspoof 2019 LA from Kaggle...")
print("Expected: ~7.6 GB, ~3-5 minutes\n")
start = time.time()

!kaggle datasets download -d anishsarkar22/asvpoof-2019-dataset-la -p {DOWNLOAD_DIR} --unzip --force

elapsed_min = (time.time() - start) / 60
print(f"\nDownload+unzip done in {elapsed_min:.1f} minutes.")

# Verify on local disk
local_la = f'{DOWNLOAD_DIR}/LA'
print(f"\nLocal LA exists: {os.path.exists(local_la)}")
if os.path.exists(local_la):
    print(f"Contents: {sorted(os.listdir(local_la))}")

Expected: ~7.6 GB, ~3-5 minutes

Dataset URL: https://www.kaggle.com/datasets/anishsarkar22/asvpoof-2019-dataset-la
License(s): ODC Attribution License (ODC-By)
100% 7.12G/7.12G [07:12<00:00, 17.7MB/s]


Download+unzip done in 8.6 minutes.

Local LA exists: True
Contents: ['ASVspoof2019_LA_asv_protocols', 'ASVspoof2019_LA_asv_scores', 'ASVspoof2019_LA_cm_protocols', 'ASVspoof2019_LA_dev', 'ASVspoof2019_LA_eval', 'ASVspoof2019_LA_train', 'README.LA.txt']


In [ ]:
import os, time

SRC = '/content/kaggle_download/LA'
DST = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'

# Make sure parent exists
os.makedirs(os.path.dirname(DST), exist_ok=True)

# Remove any partial DST from before
if os.path.exists(DST):
    print(f"Cleaning empty/stale {DST}...")
    import shutil
    shutil.rmtree(DST)

print(f"Copying {SRC} -> {DST}")
print("This time we COPY (not move) so we keep a local backup until Drive verification passes.\n")

start = time.time()
!cp -r {SRC} {DST}
elapsed = (time.time() - start) / 60
print(f"\nLocal copy done in {elapsed:.1f} minutes.")

Cleaning empty/stale /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA...
Copying /content/kaggle_download/LA -> /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA
This time we COPY (not move) so we keep a local backup until Drive verification passes.


Local copy done in 52.6 minutes.


In [ ]:
import os, glob

DST = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'

print(f"Drive copy at: {DST}")
print(f"Exists: {os.path.exists(DST)}\n")

if os.path.exists(DST):
    contents = sorted(os.listdir(DST))
    print(f"Top-level ({len(contents)} items): {contents}\n")

    expected = {'train': 25380, 'dev': 24986, 'eval': 71933}
    all_ok = True
    for p, n_expected in expected.items():
        flac_dir = os.path.join(DST, f'ASVspoof2019_LA_{p}', 'flac')
        if os.path.exists(flac_dir):
            n = len(glob.glob(os.path.join(flac_dir, '*.flac')))
            status = "OK" if n == n_expected else f"SHORT (expected {n_expected})"
            if n != n_expected:
                all_ok = False
            print(f"  [{status}] {p}: {n:,} / {n_expected:,} files")
        else:
            print(f"  [MISSING] {p}")
            all_ok = False

    print(f"\nProtocol files:")
    proto_dir = os.path.join(DST, 'ASVspoof2019_LA_cm_protocols')
    if os.path.exists(proto_dir):
        for f in sorted(os.listdir(proto_dir)):
            print(f"  - {f}")
    else:
        all_ok = False

    print(f"\nOverall: {'ALL FILES PRESENT' if all_ok else 'INCOMPLETE'}")


Drive copy at: /content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA
Exists: True

Top-level (7 items): ['ASVspoof2019_LA_asv_protocols', 'ASVspoof2019_LA_asv_scores', 'ASVspoof2019_LA_cm_protocols', 'ASVspoof2019_LA_dev', 'ASVspoof2019_LA_eval', 'ASVspoof2019_LA_train', 'README.LA.txt']

  [OK] train: 25,380 / 25,380 files
  [OK] dev: 24,986 / 24,986 files
  [OK] eval: 71,933 / 71,933 files

Protocol files:
  - ASVspoof2019.LA.cm.dev.trl.txt
  - ASVspoof2019.LA.cm.eval.trl.txt
  - ASVspoof2019.LA.cm.train.trn.txt

Overall: ALL FILES PRESENT


In [ ]:
import os
from google.colab import drive

# If something's already mounted, force-remount cleanly
try:
    drive.flush_and_unmount()
except:
    pass

drive.mount('/content/drive', force_remount=True)

# Now verify
DST = '/content/drive/MyDrive/deepfake_audio/data/raw/asvspoof_2019/LA'
print(f"\nDST exists: {os.path.exists(DST)}")

if os.path.exists(DST):
    import glob
    expected = {'train': 25380, 'dev': 24986, 'eval': 71933}
    all_ok = True
    for p, n_expected in expected.items():
        flac_dir = os.path.join(DST, f'ASVspoof2019_LA_{p}', 'flac')
        n = len(glob.glob(os.path.join(flac_dir, '*.flac'))) if os.path.exists(flac_dir) else 0
        status = "OK" if n == n_expected else "SHORT"
        if n != n_expected:
            all_ok = False
        print(f"  [{status}] {p}: {n:,} / {n_expected:,}")
    print(f"\nVerdict: {'PERSISTED CORRECTLY' if all_ok else 'DATA LOST AGAIN'}")

Drive not mounted, so nothing to flush and unmount.
Mounted at /content/drive

DST exists: True
  [SHORT] train: 0 / 25,380
  [SHORT] dev: 11,009 / 24,986
  [SHORT] eval: 0 / 71,933

Verdict: DATA LOST AGAIN


In [ ]:
import os, glob

LOCAL_LA = '/content/kaggle_download/LA'
print(f"Local copy at: {LOCAL_LA}")
print(f"Exists: {os.path.exists(LOCAL_LA)}\n")

if os.path.exists(LOCAL_LA):
    expected = {'train': 25380, 'dev': 24986, 'eval': 71933}
    all_ok = True
    for p, n_expected in expected.items():
        flac_dir = os.path.join(LOCAL_LA, f'ASVspoof2019_LA_{p}', 'flac')
        n = len(glob.glob(os.path.join(flac_dir, '*.flac'))) if os.path.exists(flac_dir) else 0
        status = "OK" if n == n_expected else "SHORT"
        if n != n_expected:
            all_ok = False
        print(f"  [{status}] {p}: {n:,} / {n_expected:,}")
    print(f"\nLocal copy: {'INTACT' if all_ok else 'INCOMPLETE'}")
else:
    print("Local copy is gone too — Colab disk was wiped on disconnect.")
    print("We'll need to re-download.")

Local copy at: /content/kaggle_download/LA
Exists: False

Local copy is gone too — Colab disk was wiped on disconnect.
We'll need to re-download.


In [ ]:
from google.colab import files
print("Re-uploading kaggle.json...")
uploaded = files.upload()

import os, shutil
os.makedirs('/root/.kaggle', exist_ok=True)
shutil.move('kaggle.json', '/root/.kaggle/kaggle.json')
os.chmod('/root/.kaggle/kaggle.json', 0o600)
print("kaggle.json configured.")

Re-uploading kaggle.json...


Saving kaggle.json to kaggle.json
kaggle.json configured.


In [ ]:
"""
SESSION BOOTSTRAP
Run this once at the start of every Colab session.
Restores: Drive mount, dataset on local disk, repo clone, Python imports.
"""
import os, sys, time

# ----- 1. Mount Drive (only used for checkpoints, NOT dataset) -----
print("=" * 60)
print("Step 1/4: Mount Drive")
print("=" * 60)
DRIVE_ROOT = '/content/drive/MyDrive'
if not os.path.exists(DRIVE_ROOT):
    from google.colab import drive
    drive.mount('/content/drive')
print("Drive mounted.\n")

# Make sure checkpoint folder exists
os.makedirs('/content/drive/MyDrive/deepfake_audio/checkpoints', exist_ok=True)
os.makedirs('/content/drive/MyDrive/deepfake_audio/logs', exist_ok=True)


# ----- 2. Clone repo to local disk -----
print("=" * 60)
print("Step 2/4: Clone repo")
print("=" * 60)
REPO_DIR = '/content/deepfake-audio-detection'
if not os.path.exists(REPO_DIR):
    !git clone https://github.com/Saracasm/deepfake-audio-detection.git {REPO_DIR}
else:
    !cd {REPO_DIR} && git pull --quiet
print(f"Repo at: {REPO_DIR}\n")


# ----- 3. Re-download dataset to local disk -----
print("=" * 60)
print("Step 3/4: Re-download dataset (~5-8 min)")
print("=" * 60)
LOCAL_LA = '/content/kaggle_download/LA'

if os.path.exists(LOCAL_LA):
    print("Dataset already present on local disk — skipping download.")
else:
    # Need kaggle.json for this — assume it's been uploaded to /root/.kaggle
    if not os.path.exists('/root/.kaggle/kaggle.json'):
        print("ERROR: kaggle.json not configured.")
        print("Run the kaggle.json upload cell first, then re-run this.")
        raise SystemExit("Need kaggle credentials")

    !pip install -q kaggle
    os.makedirs('/content/kaggle_download', exist_ok=True)

    start = time.time()
    !kaggle datasets download -d anishsarkar22/asvpoof-2019-dataset-la \
        -p /content/kaggle_download --unzip --force --quiet
    print(f"Downloaded in {(time.time()-start)/60:.1f} minutes.")

print(f"Dataset at: {LOCAL_LA}\n")


# ----- 4. Set up Python imports -----
print("=" * 60)
print("Step 4/4: Set up Python imports")
print("=" * 60)
sys.path.insert(0, REPO_DIR)

# Define key paths used throughout the notebook
LA_ROOT = LOCAL_LA  # IMPORTANT: dataset lives on LOCAL disk now, not Drive
print(f"LA_ROOT = {LA_ROOT}")
print(f"REPO_DIR = {REPO_DIR}")
print("\nBootstrap complete. Ready to work.")

Step 1/4: Mount Drive
Drive mounted.

Step 2/4: Clone repo
Cloning into '/content/deepfake-audio-detection'...
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (35/35), done.
Receiving objects: 100% (42/42), 20.42 KiB | 3.40 MiB/s, done.
remote: Total 42 (delta 13), reused 7 (delta 1), pack-reused 0 (from 0)
Resolving deltas: 100% (13/13), done.
Repo at: /content/deepfake-audio-detection

Step 3/4: Re-download dataset (~5-8 min)
Dataset URL: https://www.kaggle.com/datasets/anishsarkar22/asvpoof-2019-dataset-la
License(s): ODC Attribution License (ODC-By)
Downloaded in 3.4 minutes.
Dataset at: /content/kaggle_download/LA

Step 4/4: Set up Python imports
LA_ROOT = /content/kaggle_download/LA
REPO_DIR = /content/deepfake-audio-detection

Bootstrap complete. Ready to work.


In [ ]:
import os

REPO_DIR = '/content/deepfake-audio-detection'
files_to_check = [
    'src/data/protocols.py',
    'src/data/preprocessing.py',
    'src/data/dataset.py',
]
for f in files_to_check:
    full = os.path.join(REPO_DIR, f)
    exists = os.path.exists(full)
    size = os.path.getsize(full) if exists else 0
    print(f"  [{'OK' if exists else 'MISSING'}]  {f}  ({size} bytes)")

  [OK]  src/data/protocols.py  (3284 bytes)
  [MISSING]  src/data/preprocessing.py  (0 bytes)
  [MISSING]  src/data/dataset.py  (0 bytes)


In [ ]:
import os

# ----- preprocessing.py -----
PREPROCESSING_PY = '''"""
Audio preprocessing for ASVspoof 2019 LA.

Given a .flac path, produces a list of fixed-length 4-second segments at 16kHz
mono, ready to feed into Wav2Vec 2.0.

Pipeline:
    load .flac -> ensure mono -> ensure 16kHz -> window with 50% overlap
    (short clips are zero-padded to one full window)
"""

from typing import List
import torch
import torchaudio
import torchaudio.functional as F


SAMPLE_RATE = 16000
WINDOW_SECONDS = 4.0
OVERLAP_RATIO = 0.5

WINDOW_SAMPLES = int(SAMPLE_RATE * WINDOW_SECONDS)              # 64000
HOP_SAMPLES = int(WINDOW_SAMPLES * (1.0 - OVERLAP_RATIO))       # 32000


def load_audio(path: str, target_sr: int = SAMPLE_RATE) -> torch.Tensor:
    """Load a .flac file and return a 1-D mono waveform at target_sr."""
    waveform, sr = torchaudio.load(path)
    if waveform.shape[0] > 1:
        waveform = waveform.mean(dim=0, keepdim=True)
    if sr != target_sr:
        waveform = F.resample(waveform, orig_freq=sr, new_freq=target_sr)
    return waveform.squeeze(0)


def segment_waveform(
    waveform: torch.Tensor,
    window_samples: int = WINDOW_SAMPLES,
    hop_samples: int = HOP_SAMPLES,
) -> List[torch.Tensor]:
    """Split a 1-D waveform into fixed-length windows with overlap."""
    n = waveform.shape[0]
    if n <= window_samples:
        padded = torch.zeros(window_samples, dtype=waveform.dtype)
        padded[:n] = waveform
        return [padded]

    windows = []
    start = 0
    while start < n:
        end = start + window_samples
        if end <= n:
            windows.append(waveform[start:end])
        else:
            tail = waveform[start:]
            padded = torch.zeros(window_samples, dtype=waveform.dtype)
            padded[:tail.shape[0]] = tail
            windows.append(padded)
            break
        start += hop_samples
    return windows


def preprocess(path: str) -> List[torch.Tensor]:
    """Full pipeline: load + window."""
    waveform = load_audio(path)
    return segment_waveform(waveform)
'''

# ----- dataset.py -----
DATASET_PY = '''"""
PyTorch Dataset for ASVspoof 2019 LA.

One Dataset entry = one 4-second audio window + its binary label.
Long utterances expand into multiple entries (one per window).
"""

from typing import List, Tuple
import torch
from torch.utils.data import Dataset

from src.data.protocols import Utterance
from src.data.preprocessing import (
    load_audio,
    segment_waveform,
    WINDOW_SAMPLES,
    HOP_SAMPLES,
)


def _count_windows(duration_samples: int) -> int:
    if duration_samples <= WINDOW_SAMPLES:
        return 1
    n = (duration_samples - WINDOW_SAMPLES + HOP_SAMPLES - 1) // HOP_SAMPLES + 1
    return max(1, n)


class ASVspoofDataset(Dataset):
    """One sample = one 4-sec window + its label."""

    def __init__(
        self,
        utterances: List[Utterance],
        durations_samples: List[int] = None,
    ):
        self.utterances = utterances
        self.index: List[Tuple[int, int]] = []
        if durations_samples is None:
            for i in range(len(utterances)):
                self.index.append((i, 0))
        else:
            assert len(durations_samples) == len(utterances)
            for i, d in enumerate(durations_samples):
                n_windows = _count_windows(d)
                for w in range(n_windows):
                    self.index.append((i, w))

    def __len__(self) -> int:
        return len(self.index)

    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, int, str]:
        utt_idx, window_idx = self.index[idx]
        utt = self.utterances[utt_idx]
        waveform = load_audio(utt.flac_path)
        windows = segment_waveform(waveform)
        window_idx = min(window_idx, len(windows) - 1)
        return windows[window_idx], utt.label_int, utt.utterance_id
'''

# Write both files
files_to_write = {
    '/content/deepfake-audio-detection/src/data/preprocessing.py': PREPROCESSING_PY,
    '/content/deepfake-audio-detection/src/data/dataset.py': DATASET_PY,
}

for path, content in files_to_write.items():
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, 'w') as f:
        f.write(content)
    print(f"Wrote {path} ({len(content)} bytes)")

# Verify
print("\nFinal verification:")
for f in [
    'src/data/protocols.py',
    'src/data/preprocessing.py',
    'src/data/dataset.py',
]:
    full = os.path.join('/content/deepfake-audio-detection', f)
    size = os.path.getsize(full) if os.path.exists(full) else 0
    print(f"  [{'OK' if size > 0 else 'MISSING'}]  {f}  ({size} bytes)")

Wrote /content/deepfake-audio-detection/src/data/preprocessing.py (1990 bytes)
Wrote /content/deepfake-audio-detection/src/data/dataset.py (1741 bytes)

Final verification:
  [OK]  src/data/protocols.py  (3284 bytes)
  [OK]  src/data/preprocessing.py  (1990 bytes)
  [OK]  src/data/dataset.py  (1741 bytes)


In [ ]:
from google.colab import userdata
import os

# Get token from Secrets
GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
GITHUB_USER = "Saracasm"
REPO = "deepfake-audio-detection"

REPO_DIR = '/content/deepfake-audio-detection'
os.chdir(REPO_DIR)

# Configure git identity (needs to be re-set after each repo re-clone)
!git config user.email "95262824+Saracasm@users.noreply.github.com"
!git config user.name "Sara Iqbal"

# Stage the new files
!git add src/data/preprocessing.py src/data/dataset.py
!git status

# Commit
!git commit -m "Phase 2: add preprocessing and dataset modules"

# Push using token from secrets
push_url = f"https://{GITHUB_USER}:{GITHUB_TOKEN}@github.com/{GITHUB_USER}/{REPO}.git"
!git push {push_url} main 2>&1 | grep -v {GITHUB_TOKEN}

On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   src/data/dataset.py
	new file:   src/data/preprocessing.py

[main 1274341] Phase 2: add preprocessing and dataset modules
 2 files changed, 124 insertions(+)
 create mode 100644 src/data/dataset.py
 create mode 100644 src/data/preprocessing.py
To https://github.com/Saracasm/deepfake-audio-detection.git
   a14fd25..1274341  main -> main


In [ ]:
# from google.colab import userdata
# import os, shutil, glob

# GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
# GITHUB_USER = "Saracasm"
# REPO = "deepfake-audio-detection"
# REPO_DIR = '/content/deepfake-audio-detection'

# # Find the notebook in Drive (search for any 01_*.ipynb file)
# print("Searching for the notebook in Drive...")
# candidates = []
# for root, dirs, files in os.walk('/content/drive/MyDrive'):
#     if 'deepfake_audio' in root and 'data' in root:
#         dirs.clear()
#         continue
#     for f in files:
#         if f.endswith('.ipynb') and ('01' in f or 'data_acquisition' in f.lower()):
#             full = os.path.join(root, f)
#             candidates.append((full, os.path.getsize(full)))

# if not candidates:
#     print("\nNo notebook found. You need to first do File -> Save in Colab.")
#     print("Then re-run this cell.")
# else:
#     print(f"\nFound {len(candidates)} candidate(s):")
#     for path, size in candidates:
#         print(f"  [{size/1024:.1f} KB]  {path}")

#     # Pick the largest (most recently saved with most content, usually)
#     best = max(candidates, key=lambda x: x[1])
#     src_path = best[0]
#     print(f"\nUsing: {src_path}")

#     # Copy to repo
#     target = f"{REPO_DIR}/notebooks/01_data_acquisition.ipynb"
#     os.makedirs(os.path.dirname(target), exist_ok=True)
#     shutil.copy(src_path, target)
#     print(f"Copied to: {target}")

#     # Commit and push
#     os.chdir(REPO_DIR)
#     !git add notebooks/01_data_acquisition.ipynb
#     !git commit -m "Phase 2: add data acquisition notebook"

#     push_url = f"https://{GITHUB_USER}:{GITHUB_TOKEN}@github.com/{GITHUB_USER}/{REPO}.git"
#     !git push {push_url} main 2>&1 | grep -v {GITHUB_TOKEN}

Searching for the notebook in Drive...

Found 2 candidate(s):
  [296.7 KB]  /content/drive/MyDrive/Colab Notebooks/01_data_acquisition.ipynb
  [354.5 KB]  /content/drive/MyDrive/Colab Notebooks/Copy of 01_data_acquisition.ipynb

Using: /content/drive/MyDrive/Colab Notebooks/Copy of 01_data_acquisition.ipynb
Copied to: /content/deepfake-audio-detection/notebooks/01_data_acquisition.ipynb
[main 015ead7] Phase 2: add data acquisition notebook
 1 file changed, 1 insertion(+)
 create mode 100644 notebooks/01_data_acquisition.ipynb
remote: error: GH013: Repository rule violations found for refs/heads/main.        
remote: 
remote: - GITHUB PUSH PROTECTION        
remote:   —————————————————————————————————————————        
remote:     Resolve the following violations before pushing again        
remote: 
remote:     - Push cannot contain secrets        
remote: 
remote:             
remote:      (?) Learn how to resolve a blocked push        
remote:      https://docs.github.com/code-security

In [ ]:
import json, re

NOTEBOOK = '/content/deepfake-audio-detection/notebooks/01_data_acquisition.ipynb'

with open(NOTEBOOK, 'r') as f:
    nb = json.load(f)

# Look for any token strings in cell sources and outputs
TOKEN_PATTERN = re.compile(r'ghp_[A-Za-z0-9]{36}')
n_replaced = 0

for cell in nb.get('cells', []):
    # Clean source
    if 'source' in cell:
        if isinstance(cell['source'], list):
            for i, line in enumerate(cell['source']):
                if TOKEN_PATTERN.search(line):
                    cell['source'][i] = TOKEN_PATTERN.sub('REDACTED_TOKEN', line)
                    n_replaced += 1
        elif isinstance(cell['source'], str):
            if TOKEN_PATTERN.search(cell['source']):
                cell['source'] = TOKEN_PATTERN.sub('REDACTED_TOKEN', cell['source'])
                n_replaced += 1

    # Clean outputs (in case it appeared in command output too)
    for output in cell.get('outputs', []):
        if 'text' in output:
            if isinstance(output['text'], list):
                for i, line in enumerate(output['text']):
                    if TOKEN_PATTERN.search(line):
                        output['text'][i] = TOKEN_PATTERN.sub('REDACTED_TOKEN', line)
                        n_replaced += 1
            elif isinstance(output['text'], str):
                if TOKEN_PATTERN.search(output['text']):
                    output['text'] = TOKEN_PATTERN.sub('REDACTED_TOKEN', output['text'])
                    n_replaced += 1

with open(NOTEBOOK, 'w') as f:
    json.dump(nb, f, indent=1)

print(f"Replaced {n_replaced} occurrence(s) of the token in {NOTEBOOK}")

Replaced 2 occurrence(s) of the token in /content/deepfake-audio-detection/notebooks/01_data_acquisition.ipynb


In [ ]:
import os
os.chdir('/content/deepfake-audio-detection')

# Roll back the failed commit (but keep the file changes)
!git reset --soft HEAD~1

# Verify
!git status
!git log --oneline -3

On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   notebooks/01_data_acquisition.ipynb

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   notebooks/01_data_acquisition.ipynb

1274341 (HEAD -> main) Phase 2: add preprocessing and dataset modules
a14fd25 (origin/main, origin/HEAD) Phase 2: add ASVspoof 2019 LA protocol parser
d9f77a5 Fix in config.yaml for wandb entity


In [ ]:
import os
os.chdir('/content/deepfake-audio-detection')

!git add notebooks/01_data_acquisition.ipynb
!git status

On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   notebooks/01_data_acquisition.ipynb



In [ ]:
import re
TOKEN_PATTERN = re.compile(r'ghp_[A-Za-z0-9]{36}')

with open('notebooks/01_data_acquisition.ipynb', 'r') as f:
    content = f.read()

matches = TOKEN_PATTERN.findall(content)
print(f"Token instances remaining in notebook: {len(matches)}")
if matches:
    print("STILL CONTAINS TOKENS — do not push!")
else:
    print("Notebook is clean. Safe to push.")

Token instances remaining in notebook: 0
Notebook is clean. Safe to push.


In [ ]:
from google.colab import userdata

GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')

!git commit -m "Phase 2: add data acquisition notebook"

push_url = f"https://Saracasm:{GITHUB_TOKEN}@github.com/Saracasm/deepfake-audio-detection.git"
!git push {push_url} main 2>&1 | grep -v {GITHUB_TOKEN}

[main a3d84a6] Phase 2: add data acquisition notebook
 1 file changed, 4486 insertions(+)
 create mode 100644 notebooks/01_data_acquisition.ipynb
To https://github.com/Saracasm/deepfake-audio-detection.git
   1274341..a3d84a6  main -> main


In [ ]:
import sys, importlib

# Make sure repo is on Python path
sys.path.insert(0, '/content/deepfake-audio-detection')

# Reload modules in case anything was cached
for mod in ['src.data.protocols', 'src.data.preprocessing', 'src.data.dataset']:
    if mod in sys.modules:
        importlib.reload(sys.modules[mod])

from src.data.protocols import parse_all_partitions, class_counts
from src.data.preprocessing import load_audio, segment_waveform, WINDOW_SAMPLES, HOP_SAMPLES
from src.data.dataset import ASVspoofDataset, _count_windows

# Re-parse protocols against the local dataset
LA_ROOT = '/content/kaggle_download/LA'
splits = parse_all_partitions(LA_ROOT)

print("Modules imported, protocols parsed:")
for name, utts in splits.items():
    counts = class_counts(utts)
    print(f"  {name}: {len(utts):,} ({counts['bonafide']:,} bonafide, {counts['spoof']:,} spoof)")

Modules imported, protocols parsed:
  train: 25,380 (2,580 bonafide, 22,800 spoof)
  dev: 24,844 (2,548 bonafide, 22,296 spoof)
  eval: 71,237 (7,355 bonafide, 63,882 spoof)


In [15]:
import torchaudio
from tqdm import tqdm

print("Measuring durations of all training clips...")

# Try torchaudio.info first (fast, metadata only); fall back to load if needed
try:
    sample = splits['train'][0]
    test = torchaudio.info(sample.flac_path)
    use_info = True
    print("Using fast metadata reads (torchaudio.info)\n")
except (AttributeError, RuntimeError):
    use_info = False
    print("Falling back to torchaudio.load (slower but works)\n")

train_durations = []
for u in tqdm(splits['train'], desc="Train"):
    if use_info:
        train_durations.append(torchaudio.info(u.flac_path).num_frames)
    else:
        waveform, _ = torchaudio.load(u.flac_path)
        train_durations.append(waveform.shape[1])

print(f"\nMeasured {len(train_durations):,} clip durations.")
print(f"Min:    {min(train_durations):>10,} samples ({min(train_durations)/16000:.2f} sec)")
print(f"Max:    {max(train_durations):>10,} samples ({max(train_durations)/16000:.2f} sec)")
print(f"Mean:   {sum(train_durations)/len(train_durations):>10,.0f} samples ({sum(train_durations)/len(train_durations)/16000:.2f} sec)")

Measuring durations of all training clips...
Falling back to torchaudio.load (slower but works)



Train: 100%|██████████| 25380/25380 [01:06<00:00, 382.25it/s]


Measured 25,380 clip durations.
Min:        10,439 samples (0.65 sec)
Max:       211,007 samples (13.19 sec)
Mean:       54,813 samples (3.43 sec)


In [16]:
# Build the training Dataset
train_dataset = ASVspoofDataset(
    utterances=splits['train'],
    durations_samples=train_durations,
)

n_utterances = len(splits['train'])
n_windows = len(train_dataset)
inflation = n_windows / n_utterances

print(f"Training set:")
print(f"  Utterances:       {n_utterances:,}")
print(f"  Total windows:    {n_windows:,}")
print(f"  Inflation factor: {inflation:.2f}x")
print(f"  (i.e., on average {inflation:.2f} windows per utterance)")

# Spot-check a few entries from the index
print(f"\nFirst 5 entries from the dataset index:")
for i in range(5):
    utt_idx, window_idx = train_dataset.index[i]
    utt = train_dataset.utterances[utt_idx]
    duration = train_durations[utt_idx]
    n_w_for_utt = sum(1 for j in train_dataset.index if j[0] == utt_idx)
    print(f"  idx {i}: utt={utt.utterance_id} ({duration} samples / {duration/16000:.2f} sec), "
          f"window {window_idx} of {n_w_for_utt}")

Training set:
  Utterances:       25,380
  Total windows:    34,036
  Inflation factor: 1.34x
  (i.e., on average 1.34 windows per utterance)

First 5 entries from the dataset index:
  idx 0: utt=LA_T_1138215 (55329 samples / 3.46 sec), window 0 of 1
  idx 1: utt=LA_T_1271820 (70323 samples / 4.40 sec), window 0 of 2
  idx 2: utt=LA_T_1271820 (70323 samples / 4.40 sec), window 1 of 2
  idx 3: utt=LA_T_1272637 (46392 samples / 2.90 sec), window 0 of 1
  idx 4: utt=LA_T_1276960 (45001 samples / 2.81 sec), window 0 of 1


Inflation factor 1.34x — meaning training will see 34,036 samples per epoch instead of 25,380. That extra 8,656 windows comes from longer clips contributing multiple windows. Within the range we expected (1.1-1.5).The index walks correctly:

idx 0: a 3.46-sec clip → 1 window (window 0 of 1)
idx 1: a 4.40-sec clip → 2 windows, this is window 0
idx 2: same clip → window 1 of 2 (the overlapping window covering 2-6 sec)
idx 3-4: short clips → 1 window each

In [17]:
import torch
from torch.utils.data import DataLoader

# Build a DataLoader with a small batch and a few worker processes
train_loader = DataLoader(
    train_dataset,
    batch_size=8,
    shuffle=True,
    num_workers=2,        # parallel data loading
    pin_memory=True,      # faster GPU transfer
)

print(f"DataLoader built. Pulling one batch...")

# Pull exactly one batch
batch_iter = iter(train_loader)
waveforms, labels, utt_ids = next(batch_iter)

print(f"\nBatch shapes:")
print(f"  waveforms: {waveforms.shape}  (batch_size x samples)")
print(f"  labels:    {labels.shape}     (batch_size,)")
print(f"  utt_ids:   list of {len(utt_ids)} strings")

print(f"\nBatch contents:")
print(f"  waveform dtype:  {waveforms.dtype}")
print(f"  label values:    {labels.tolist()}")
print(f"  utt_ids:         {list(utt_ids)}")

print(f"\nWaveform statistics:")
print(f"  Min:  {waveforms.min().item():.3f}")
print(f"  Max:  {waveforms.max().item():.3f}")
print(f"  Mean: {waveforms.mean().item():.4f}")

DataLoader built. Pulling one batch...

Batch shapes:
  waveforms: torch.Size([8, 64000])  (batch_size x samples)
  labels:    torch.Size([8])     (batch_size,)
  utt_ids:   list of 8 strings

Batch contents:
  waveform dtype:  torch.float32
  label values:    [1, 1, 1, 1, 1, 0, 1, 1]
  utt_ids:         ['LA_T_7084517', 'LA_T_3950907', 'LA_T_5280537', 'LA_T_4465139', 'LA_T_5033641', 'LA_T_3402710', 'LA_T_9651282', 'LA_T_1892752']

Waveform statistics:
  Min:  -1.000
  Max:  1.000
  Mean: -0.0000
